### PR TITLE
kv API update to get predicate when needed for SetIf operation

### LINKED ISSUE #3400: kv API predicate handling
The current API to `SetIf` enable the developer to pass a value as the predicate ([]byte).
To enable any underlaying implementation to use `SetIf` with any value of a predicate - returned from a `Get` operation, we should design the API to same.

The Get should return an opaque value (interface{}) that should be passed to the `SetIf` operation. Using this the implementor and the developer using the API will be a clear way to implement and pass the predicate value.

### PR DESCRIPTION
**Pull request by nopcoder**
Close #3400

### CODE REVIEW

**Review by itaiad200 — CHANGES_REQUESTED** [review]
I think we should replace the `Get` API and not add another one

File: pkg/kv/store.go [hunk 1]
```diff
@@ -41,18 +41,22 @@ type Driver interface {
...
 }
 
+type Predicate interface{}
+
 type Store interface {
 	// Get returns a value for the given key, or ErrNotFound if key doesn't exist
 	Get(ctx context.Context, key []byte) ([]byte, error)
 
+	GetEntry(ctx context.Context, key []byte) (*Entry, error)
+
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply pkg/kv/store.go]: Renamed to GetValuaPredicate - the case where Get without predicate is a lighter version of Get with predicate.
I've tested the "options" variant, where we have one Get and one Set method and passing options to the function control if you like to pass a predicate or not.
The problem I see in this style is the contract that the implementor need to address - the combination of all the options and the expected result.
When you have a method signature - it is easy to pass the requirement for the implementor.

File: pkg/kv/store_message.go [hunk 2]
```diff
@@ -13,55 +12,39 @@ type StoreMessage struct {
...
-func (s *StoreMessage) GetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
-	val, err := s.Store.Get(ctx, []byte(path))
+func (s *StoreMessage) GetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) (Predicate, error) {
+	ent, err := s.Store.GetEntry(ctx, []byte(path))
 	if err != nil {
-		return fmt.Errorf("failed on Get (path: %s): %w", path, err)
+		return nil, err
 	}
-	return proto.Unmarshal(val, msg)
+	if msg == nil {
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply pkg/kv/store_message.go:28]: It was a quick way to get a predicate for a key without serialize the value.

File: pkg/kv/store.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (itaiad200) [reply pkg/kv/store.go]: I get why we would like to have 2 `Set`s functions, since the predicate existence changes the behaviour.
I don't get why there's a need for 2 `Get`s. We pass the predicate and if the caller wish, she can ignore it. It's not saving any time/memory, just makes the api more confusing..

File: pkg/kv/store_message.go [hunk 3]
```diff
@@ -14,54 +13,46 @@ type StoreMessage struct {
 }
 
 func (s *StoreMessage) GetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
-	val, err := s.Store.Get(ctx, []byte(path))
+	v, err := s.Store.Get(ctx, []byte(path))
 	if err != nil {
-		return fmt.Errorf("failed on Get (path: %s): %w", path, err)
+		return err
```
Comment (N-o-Z) [inline pkg/kv/store_message.go:18]: Why did you remove the formatting? It was the only place where we log the key that failed

File: pkg/kv/store_message.go [hunk 4]
```diff
@@ -14,54 +13,46 @@ type StoreMessage struct {
...
+	return proto.Unmarshal(v, msg)
 }
 
-func (s *StoreMessage) SetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
-	val, err := proto.Marshal(msg)
+func (s *StoreMessage) GetMsgPredicate(ctx context.Context, path string, msg protoreflect.ProtoMessage) (Predicate, error) {
+	vp, err := s.Store.GetValuePredicate(ctx, []byte(path))
 	if err != nil {
-		return fmt.Errorf("failed on Set (path: %s): %w", path, err)
+		return nil, err
```
Comment (N-o-Z) [inline pkg/kv/store_message.go:26]: Same as before

File: pkg/kv/store_message.go [hunk 5]
```diff
@@ -14,54 +13,46 @@ type StoreMessage struct {
...
 
-func (s *StoreMessage) SetIf(ctx context.Context, path string, msg protoreflect.ProtoMessage, pred protoreflect.ProtoMessage) error {
+func (s *StoreMessage) SetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
 	val, err := proto.Marshal(msg)
 	if err != nil {
-		return fmt.Errorf("failed on Marshal (path: %s): %w", path, err)
-	}
-	if pred == nil {
-		return s.Store.SetIf(ctx, []byte(path), val, nil)
+		return err
```
Comment (N-o-Z) [inline pkg/kv/store_message.go:41]: Same as previous comment

File: pkg/kv/store_message.go [hunk 6]
```diff
@@ -14,54 +13,46 @@ type StoreMessage struct {
...
+	return s.Store.SetIf(ctx, []byte(path), val, predicate)
 }
 
 func (s *StoreMessage) Delete(ctx context.Context, path string) error {
-	err := s.Store.Delete(ctx, []byte(path))
-	if err != nil {
-		err = fmt.Errorf("failed on Delete (path: %s): %w", path, err)
-	}
-	return err
+	return s.Store.Delete(ctx, []byte(path))
```
Comment (N-o-Z) [inline pkg/kv/store_message.go:55]: The error was here for a reason - this changes the multiparts tracker current behavior

File: pkg/kv/store.go [hunk 7]
```diff
@@ -41,18 +41,22 @@ type Driver interface {
 	Open(ctx context.Context, dsn string) (Store, error)
 }
 
+type Predicate interface{}
+
 type Store interface {
 	// Get returns a value for the given key, or ErrNotFound if key doesn't exist
 	Get(ctx context.Context, key []byte) ([]byte, error)
 
+	GetValuePredicate(ctx context.Context, key []byte) (*ValuePredicate, error)
```
Comment (N-o-Z) [inline pkg/kv/store.go:50]: Missing documentation

File: pkg/kv/store.go [hunk 8]
```diff
@@ -81,12 +85,18 @@ type EntriesIterator interface {
 	Close()
 }
 
-// Entry holds a pair of key/value
+// Entry holds a pair of key/value and predicate used for conditional set
```
Comment (N-o-Z) [inline pkg/kv/store.go:88]: No predicate in struct

**Review by itaiad200 — CHANGES_REQUESTED** [review]
I still think we should have a single `Get` API for the store

File: pkg/kv/store_message.go (hunk 5 above)
Comment (nopcoder) [reply pkg/kv/store_message.go:41]: Will address it as part of Store implementation

Comment (N-o-Z) [reply pkg/kv/store_message.go:41]: I wonder why you have decided to move it to the store implementation, as this will require all stores to make sure to log the error with the key. When done in the store_message level you only have to take care of it once WDYT?

File: pkg/kv/store_message.go (hunk 6 above)
Comment (N-o-Z) [reply pkg/kv/store_message.go:55]: The comment was regarding the removal of the check in delete to see if the key exists and if not return an error

File: pkg/kv/store.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nopcoder) [reply pkg/kv/store.go]: Didn't want to lock the option that there can be a version of 'get' with predicate that will slower to use than a 'get' for the latest value (without predicate).
I'll discover and see how we plan to implement the 'get' for different underlying storage types - if it is the same cost as calling a simple 'get' - I'll make predicate optional and leave one API.

File: pkg/kv/store_message.go (hunk 5 above)
Comment (nopcoder) [reply pkg/kv/store_message.go:41]: Wanted to get the infomation also when you directly access the storage.
- We can guide the storage implementor
- Create a wrapper that will wrap the underlying storage to enritch the return errors
WDYT?

File: pkg/kv/store_message.go (hunk 6 above)
Comment (nopcoder) [reply pkg/kv/store_message.go:55]: The previous code - for any error in delete enrich the error with path - no extra verification at this level.
```go
        err := s.Store.Delete(ctx, []byte(path))
	if err != nil {
		err = fmt.Errorf("failed on Delete (path: %s): %w", path, err)
	}
	return err
```

Comment (N-o-Z) [reply pkg/kv/store_message.go:56]: Sorry, my bad - it's handled somewhere else

File: pkg/kv/store_message.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/kv/store_message.go:22]: Maybe we should document this functionality in the function so the user understands that passing nil msg to GetMsg is a feature :)

**Review by N-o-Z — APPROVED** [review]

File: pkg/kv/store.go [hunk 9]
```diff
@@ -41,9 +41,23 @@ type Driver interface {
 	Open(ctx context.Context, dsn string) (Store, error)
 }
 
+// Predicate value used to update a key base on a previous fetched value.
+//   Store's GetValuePredicate used to pull the key's value with the associated predicate.
```
Comment (itaiad200) [inline pkg/kv/store.go:45]:
```suggestion
//   Store's Get used to pull the key's value with the associated predicate.
```

File: pkg/kv/store.go [hunk 10]
```diff
@@ -41,9 +41,23 @@ type Driver interface {
...
+//   Store's SetIf used to set the key's value based on the predicate.
+type Predicate interface{}
+
+// GetResult value with predicate - Value holds the data and Predicate a value used for conditional set.
+//   Get operation will return this struct, holding the key's information
+//   SetIf operation will use the Predicate for conditional set
+type GetResult struct {
```
Comment (itaiad200) [inline pkg/kv/store.go:52]: Perhaps 
```suggestion
type ValueWithPredicate struct {
```

File: pkg/kv/mem/store.go [hunk 11]
```diff
@@ -91,18 +95,18 @@ func (s *Store) SetIf(_ context.Context, key, value, valuePredicate []byte) erro
 	curr, currOK := s.m[string(key)]
 	if valuePredicate == nil {
 		if currOK {
-			return kv.ErrPredicateFailed
+			return fmt.Errorf("%w: key=%v", kv.ErrPredicateFailed, key)
```
Comment (itaiad200) [inline pkg/kv/mem/store.go:98]:
```suggestion
			return fmt.Errorf("key=%v: %w", key, kv.ErrPredicateFailed)
```

Comment (itaiad200) [reply pkg/kv/mem/store.go:98]: Here and elsewhere.

**Review by itaiad200 — APPROVED** [review]
LGTM
just mind the error formating
